process/block/poolsCleaner/miniBlocksCleaner.go

```diff
@@ -0,0 +1,153 @@
+		return nil, process.ErrNilTransactionPool
+	}
+	if check.IfNil(rounder) {
+		return nil, process.ErrNilRounder
+	}
+	if check.IfNil(shardCoordinator) {
+		return nil, process.ErrNilShardCoordinator
+	}
+
+	mbpc := miniBlocksPoolsCleaner{
+		blockTracker:     blockTracker,
+		dataPool:         dataPool,
+		rounder:          rounder,
+		shardCoordinator: shardCoordinator,
+	}
+
+	mbpc.mapMiniBlocksRounds = make(map[string]int64)
+	miniBlocksPool := mbpc.dataPool.MiniBlocks()
+	miniBlocksPool.RegisterHandler(mbpc.receivedMiniBlock)
+
+	return &mbpc, nil
+}
+
+func (mbpc *miniBlocksPoolsCleaner) receivedMiniBlock(key []byte) {
+	if key == nil {
+		return
+	}
+
+	log.Trace("miniBlocksPoolsCleaner.receivedMiniBlock", "hash", key)
+
+	mbpc.mutMapMiniBlocksRounds.Lock()
+	defer mbpc.mutMapMiniBlocksRounds.Unlock()
+
+	if _, ok := mbpc.mapMiniBlocksRounds[string(key)]; !ok {
+		mbpc.mapMiniBlocksRounds[string(key)] = mbpc.rounder.Index()
+	}
+
+	mbpc.cleanPoolIfNeeded()
+}
+
+func (mbpc *miniBlocksPoolsCleaner) cleanPoolIfNeeded() {
+	selfShard := mbpc.shardCoordinator.SelfId()
+	numPendingMiniBlocks := mbpc.blockTracker.GetNumPendingMiniBlocks(selfShard)
+	miniBlocksPool := mbpc.dataPool.MiniBlocks()
+	transactionsPool := mbpc.dataPool.Transactions()
+	percentUsed := float64(miniBlocksPool.Len()) / float64(miniBlocksPool.MaxSize())
+
+	for hash, round := range mbpc.mapMiniBlocksRounds {
+		value, ok := miniBlocksPool.Get([]byte(hash))
+		if !ok {
+			log.Trace("miniblock not found in pool",
+				"hash", []byte(hash),
+				"round", round)
+			delete(mbpc.mapMiniBlocksRounds, hash)
+			continue
+		}
+
+		miniBlock, ok := value.(*block.MiniBlock)
+		if !ok {
+			log.Debug("cleanPoolIfNeeded", "error", process.ErrWrongTypeAssertion,
+				"hash", []byte(hash),
+				"round", round)
+			continue
+		}
+
+		if miniBlock.SenderShardID != selfShard {
+			if numPendingMiniBlocks > 0 && percentUsed < percentAllowed {
+				log.Trace("cleaning cross miniblock not yet allowed",
+					"hash", []byte(hash),
+					"round", round,
+					"num pending miniblocks", numPendingMiniBlocks,
+					"miniblocks pool percent used", percentUsed,
+					"type", miniBlock.Type,
+					"sender", miniBlock.SenderShardID,
+					"receiver", miniBlock.ReceiverShardID,
+					"num txs", len(miniBlock.TxHashes))
+				continue
+			}
+		}
+
+		roundDif := mbpc.rounder.Index() - round
+		if roundDif <= process.MaxRoundsToKeepUnprocessedMiniBlocks {
+			log.Trace("cleaning miniblock not yet allowed",
+				"hash", []byte(hash),
+				"round", round,
+				"round dif", roundDif,
+				"type", miniBlock.Type,
+				"sender", miniBlock.SenderShardID,
+				"receiver", miniBlock.ReceiverShardID,
+				"num txs", len(miniBlock.TxHashes))
+			continue
+		}
+
+		strCache := process.ShardCacherIdentifier(miniBlock.SenderShardID, miniBlock.ReceiverShardID)
+		transactionsPool.RemoveSetOfDataFromPool(miniBlock.TxHashes, strCache)
+		miniBlocksPool.Remove([]byte(hash))
+		delete(mbpc.mapMiniBlocksRounds, hash)
```

Comment:
If the node is currently syncing and is 200 rounds behind he will receive the miniblocks that are being broadcasted right now. The blocks that he needs first will be removed first from all pools?

Reply:
Actually both of them are not removed as they are received around the same current chronology round. The removal action is triggered if they will remain unprocessed for at least 100 rounds. The condition to be removed is not the number of the miniblocks received, but theirs age since the first time when a specific hash has been received.

Reply:
Combined with peding miniblocks protection for cross miniblocks

Reply:
However even if some of them would be not processed after 100 rounds and would be removed from pool the node would request them as soon as it would need them and they are not found in the pool

Reply:
ok